File: app/src/main/java/com/blog/demo/custom/widget/VelocityTrackerView.java
package com.blog.demo.custom.widget;

import android.content.Context;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.VelocityTracker;
import android.view.ViewConfiguration;
import android.widget.Scroller;
import android.widget.TextView;

public class VelocityTrackerView extends TextView {

    private Scroller mScroller;
    private VelocityTracker mVelocityTracker;

    private int mDownX;
    private boolean mSlide;
    private int mMaxVelocity;
    private int mTouchSlop;

    public VelocityTrackerView(Context context) {
        this(context, null);
    }

    public VelocityTrackerView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        mScroller = new Scroller(context);

        ViewConfiguration configuration = ViewConfiguration.get(context);
        mMaxVelocity = configuration.getScaledMaximumFlingVelocity();
        mTouchSlop = configuration.getScaledTouchSlop();
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        switch(ev.getAction()) {
            case MotionEvent.ACTION_DOWN:
                mDownX = (int) ev.getX();
                // 请求一个新的VelocityTracker
                acquireVelocityTracker(ev);
                break;
            case MotionEvent.ACTION_MOVE:
                if (mVelocityTracker != null) {
                    // 添加触摸对象
                    mVelocityTracker.addMovement(ev);
                    // 计算当前速率
                    mVelocityTracker.computeCurrentVelocity(1000, mMaxVelocity);
                    final float velocityX = mVelocityTracker.getXVelocity();
                    if (Math.abs(velocityX) > 600 &&
                            Math.abs(ev.getX() - mDownX) > mTouchSlop) {
                        mSlide = true;
                    }
                }
                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                if (mSlide) {
                    // 回到初始位置
                    backToOrigin(mDownX - (int) ev.getX(), 0);
                    mSlide = false;
                }
                if (mVelocityTracker != null )
                    releaseVelocityTracker();
                break;
        }
        return super.dispatchTouchEvent(ev);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch(event.getAction()) {
            case MotionEvent.ACTION_MOVE:
                if (mSlide) {
                    scrollTo(mDownX - (int) event.getX(), 0);
                }
                break;
        }
        return true;
    }

    private void acquireVelocityTracker(final MotionEvent event) {
        if (mVelocityTracker == null) {
            mVelocityTracker = VelocityTracker.obtain();
        }

        mVelocityTracker.addMovement(event);
    }

    private void releaseVelocityTracker() {
        if (mVelocityTracker != null) {
            mVelocityTracker.clear();
            mVelocityTracker.recycle();
            mVelocityTracker = null;
        }
    }

    // 移动到原点
    public void backToOrigin (int scrollX, int scrollY) {
        mScroller.startScroll(scrollX, scrollY, -scrollX, -scrollY, 1000);
        postInvalidate();
    }

    @Override
    public void computeScroll() {
        if (mScroller.computeScrollOffset()) {
            scrollTo(mScroller.getCurrX(), mScroller.getCurrY());

            postInvalidate();
        }
    }
}

File: app/src/main/java/com/blog/demo/image/widget/VelocityTrackerView.kt
package com.blog.demo.image.widget

import android.content.Context
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.VelocityTracker
import android.view.ViewConfiguration
import android.widget.Scroller
import kotlin.math.abs

class VelocityTrackerView (context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
    androidx.appcompat.widget.AppCompatTextView(context, attrs, defStyleAttr) {

    private var mScroller: Scroller
    private var mVelocityTracker: VelocityTracker? = null

    private var mDownX = 0
    private var mSlide = false
    private var mMaxVelocity = 0
    private var mTouchSlop = 0

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    init {
        mScroller = Scroller(context)

        val configuration = ViewConfiguration.get(context)
        mMaxVelocity = configuration.scaledMaximumFlingVelocity
        mTouchSlop = configuration.scaledTouchSlop
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        when (ev.action) {
            MotionEvent.ACTION_DOWN -> {
                mDownX = ev.x.toInt()
                // 请求一个新的VelocityTracker
                acquireVelocityTracker(ev)
            }
            MotionEvent.ACTION_MOVE -> {
                if (mVelocityTracker != null) {
                    // 添加触摸对象
                    mVelocityTracker?.addMovement(ev)
                    // 计算当前速率
                    mVelocityTracker?.computeCurrentVelocity(1000, mMaxVelocity.toFloat())
                    val velocityX = mVelocityTracker?.xVelocity ?: 0f
                    if (abs(velocityX) > 600 && abs(ev.x - mDownX) > mTouchSlop) {
                        mSlide = true
                    }
                }
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                if (mSlide) {
                    // 回到初始位置
                    backToOrigin(mDownX - ev.x.toInt(), 0)
                    mSlide = false
                }
                if (mVelocityTracker != null) releaseVelocityTracker()
            }
        }
        return super.dispatchTouchEvent(ev)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_MOVE -> if (mSlide) {
                scrollTo(mDownX - event.x.toInt(), 0)
            }
        }
        return true
    }

    private fun acquireVelocityTracker(event: MotionEvent) {
        if (mVelocityTracker == null) {
            mVelocityTracker = VelocityTracker.obtain()
        }
        mVelocityTracker?.addMovement(event)
    }

    private fun releaseVelocityTracker() {
        if (mVelocityTracker != null) {
            mVelocityTracker?.clear()
            mVelocityTracker?.recycle()
            mVelocityTracker = null
        }
    }

    // 移动到原点
    fun backToOrigin(scrollX: Int, scrollY: Int) {
        mScroller.startScroll(scrollX, scrollY, -scrollX, -scrollY, 1000)
        postInvalidate()
    }

    override fun computeScroll() {
        if (mScroller.computeScrollOffset()) {
            scrollTo(mScroller.currX, mScroller.currY)
            postInvalidate()
        }
    }

}